### PR TITLE
[BugFix]Fix NoSuchElementException from query hive partition table, which has no partition

### LINKED ISSUE #6613: [Bug]Query from empty partition hive table results in NoSuchElementException
### Steps to reproduce the behavior (Required)
Query with partition key as where clause, from hive table which has no partition leads to null pointer exeception
1.  CREATE EXTERNAL TABLE testpar (
 id_card int(11) NULL COMMENT "",
 tran_time varchar(65533) NULL COMMENT "",
 name varchar(65533) NULL COMMENT "",
 cash int(11) NULL COMMENT "",
 ds varchar(65533) NULL COMMENT ""
) ENGINE=HIVE 
COMMENT "PARTITION BY (ds)"
PROPERTIES (
"database" = "un_test",
"table" = "testpar",
"resource" = "hive0",
"hive.metastore.uris" = "thrift://127.0.0.1:9083");
2. Select * from testpar where ds > '1000';

### Expected behavior (Required)
MySQL [demodb]> select * from testpar where ds > '1000';
Empty set (0.01 sec)

### Real behavior (Required)
2022-05-25 14:36:04,998 WARN (starrocks-mysql-nio-pool-18|1203) [StmtExecutor.execute():324] New planner error: select * from st_cl_user_ne_app_hour where minute>'1400'
com.starrocks.sql.common.StarRocksPlannerException: null
 at com.starrocks.sql.optimizer.rule.transformation.RemoteScanPartitionPruneRule.transform(RemoteScanPartitionPruneRule.java:68) ~[starrocks-fe.jar:?] 

### StarRocks version (Required)
 - 2.2.0

### PR DESCRIPTION
**Pull request by adzfolc**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes https://github.com/StarRocks/starrocks/issues/6613

## Problem Summary(Required) ：

1. How the bug is triggered?
    * When user queries from hive partitioned table with partition key in where clause, optimizer rule ListPartitionPruner will prune unreachable partitions.
    * In ListPartitionPruner, Map<partition key, partition id> is stored in partitionValueMap.When execute ListPartitionPruner, optimizer needs to fetch partition id from partitionValueMap.
    * If external table has no partitions, then when optimizer fetching from partitionValueMap, will leads to NoSuchElementException.
2. Bug example:
   * Select * from ex_tbl_par where parKey>'1000';
   * Select * from ex_tbl_par where parKey in ('1000');
   * explanation
        * ex_tbl_par: only external partition table having no partitons is effected now
        * parKey: partition key of external partition table
        * where clause: the binary operator of where clause is one of follows:
            * LE("<=")
            * GE(">=")
            * LT("<")
            * GT(">")
        * in clause: InPredicateOperator has partiton key as conditions
3. Hot to fix
    * Before pruning partitions, check partitionValueMap.If partitionValueMap is null, then return null.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java|2|2|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java [hunk 1]
```diff
@@ -190,7 +190,7 @@ private Set<Long> evalBinaryPredicate(BinaryPredicateOperator binaryPredicate) {
         Set<Long> matches = Sets.newHashSet();
         TreeMap<LiteralExpr, Set<Long>> partitionValueMap = columnToPartitionValuesMap.get(leftChild);
         Set<Long> nullPartitions = columnToNullPartitions.get(leftChild);
-        if (partitionValueMap == null || nullPartitions == null) {
+        if (partitionValueMap == null || nullPartitions == null || partitionValueMap.size() == 0) {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java:193]: %s/size()==0/isEmpty()/g

Comment (adzfolc) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java]: > %s/size()==0/isEmpty()/g

done

**Comment by adzfolc** [comment]
> @adzfolc Plz add ut

added

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ListPartitionPruner.java|2|2|100.00%|

**Review by DorianZheng — APPROVED** [review]
LGTM

**Review by imay — APPROVED** [review]

**Review by caneGuy — APPROVED** [review]

**Comment by adzfolc** [comment]
@Mergifyio backport branch-2.2 branch-2.1 branch-2.3

**Pull request merged**